### PR TITLE
add codegen for general pointwise kernels

### PR DESCRIPTION
**Pull request by iclementine**
1. add codegen utilities for general pointwise kernels( both dynamic & static)
   - dynamic: speciallize to the rank of the task space;
   - static: speciallize to the shape & strides of each input tensor;
2. use pointwise_dynamic to replace abs, gelu, reciprocals, relu, rsqrt, silu and implement sigmoid

### CODE REVIEW

File: src/flag_gems/gelu.py [hunk 1]
```diff
@@ -3,115 +3,37 @@
 import triton.language as tl
 from .__libentry__ import libentry
```
Comment (StrongSpoon) [inline src/flag_gems/gelu.py]: this line could be deleted

File: src/flag_gems/sigmoid.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import torch
+import triton
+import triton.language as tl
+from flag_gems.utils.pointwise_dynamic import pointwise_dynamic
+
+@pointwise_dynamic
+@triton.jit
+def sigmoid_forward(x):
+    return 1 / (1 + tl.exp(-x.to(tl.float32)))
```
Comment (tongxin) [inline src/flag_gems/sigmoid.py]: Isn't it preferable to convert exp to math.exp2?

File: src/flag_gems/utils/shape_utils.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
+from typing import Tuple, Iterable
+import functools
+import operator
+
+Shape = Tuple[int]
+Stride = Tuple[int]
+MultiIndex = Tuple[int]
+Perm = Tuple[int]
+
+
```
Comment (tongxin) [inline src/flag_gems/utils/shape_utils.py:10]: This whole shape util could be slow due to Python interpreter. I would suggest calculating with numpy arrays.

File: src/flag_gems/sigmoid.py (hunk 2 above)
Comment (iclementine) [reply src/flag_gems/sigmoid.py]: Ok, done.

File: src/flag_gems/utils/shape_utils.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    d = r2 - r1
+    new_stride = [0 for _ in range(r2)]
+    for i in range(r1):
+        new_stride[d + i] = 0 if (shape[i] == 1 and new_shape[d + i] > 1) else stride[i]
+    return tuple(new_stride)
+
+def volume(shape: Shape) -> int:
+    return functools.reduce(operator.mul, shape, 1)
+
+def is_valid_perm(perm: Perm) -> bool:
```
Comment (StrongSpoon) [inline src/flag_gems/utils/shape_utils.py:95]: simplify this fn

File: src/flag_gems/utils/code_cache.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+def cache_dir_path() -> Path:
+    """Return the cache directory for generated files in flaggems."""
+    _cache_dir = os.environ.get("FLAGGEMS_CACHE_DIR")
+    if _cache_dir is None:
+        _cache_dir = Path.home() / ".flaggems"
+    else:
+        _cache_dir = Path(_cache_dir)
+    return _cache_dir
+
+@functools.lru_cache(None)
```
Comment (tongxin) [inline src/flag_gems/utils/code_cache.py]: I would suggest using functools.cached_property, for cache_dir should be invariant once initiated.

File: src/flag_gems/utils/code_cache.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+def cache_dir() -> Path:
+    """Return cache directory for generated files in flaggems. Create it if it does not exist."""
+    _cache_dir = cache_dir_path()
+    os.makedirs(_cache_dir, exist_ok=True)
+    return _cache_dir
+
+def clear_cache():
+    """Clear the cache directory for code cache."""
+    _cache_dir = cache_dir_path()
```
Comment (tongxin) [inline src/flag_gems/utils/code_cache.py:24]: then here simply querying cache_dir() could be quicker

File: src/flag_gems/utils/inliner.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    mapping = {}
+    # replace arguemnt names with input names
+    # TODO: handle non-tensor arguments to scalar function
+    for p, input_name in zip(f.arg_names, input_names):
+        mapping[p] = input_name
+
+    # replace temporaries with auto-named names to avoid conflicts
+    for name in nc.names:
+        mapping[name] = namespace.create_name(name, None)
+
```
Comment (tongxin) [inline src/flag_gems/utils/inliner.py]: There could come name conflicts with the rewriting. For example, 

foo (a, b):
   x = ...
   y = 

If rewriting arg a, b with input name x, y, it will result in name aliasing.

File: src/flag_gems/utils/shape_utils.py (hunk 4 above)
Comment (iclementine) [reply src/flag_gems/utils/shape_utils.py:95]: Done.  Now the perm is sorted and test if sorted[i] = i for all i in [0, r).

File: src/flag_gems/gelu.py (hunk 1 above)
Comment (iclementine) [reply src/flag_gems/gelu.py]: done

File: src/flag_gems/utils/code_cache.py (hunk 5 above)
Comment (iclementine) [reply src/flag_gems/utils/code_cache.py]: I use `functools.cache` since it is a standalone function.
Also `functools.cache` is more light-weight than `functools.lrucache`/

File: src/flag_gems/utils/code_cache.py (hunk 6 above)
Comment (iclementine) [reply src/flag_gems/utils/code_cache.py:24]: Done.

File: src/flag_gems/utils/inliner.py (hunk 7 above)
Comment (iclementine) [reply src/flag_gems/utils/inliner.py]: Input names to the inlined function are populated into the namespace before inlining, so in this case, `x`, `y` would be renamed to `x_1`, `y_1`.

Also, now all the args names and local names of the jit function, into which the inlined function is inlined, are populated into the namespace brefore inlining.

File: src/flag_gems/utils/shape_utils.py (hunk 4 above)
Comment (iclementine) [reply src/flag_gems/utils/shape_utils.py:95]: done

File: src/flag_gems/gelu.py (hunk 1 above)
Comment (iclementine) [reply src/flag_gems/gelu.py]: done

File: src/flag_gems/utils/shape_utils.py (hunk 3 above)
Comment (iclementine) [reply src/flag_gems/utils/shape_utils.py:10]: Some operations used in shape_utils are not simple array operation. So I prefer using python tuples & list. We  may find a better implementation after benchmarking them.
